src/App.tsx

```diff
@@ -2,24 +1,66 @@
 import './App.css';
-import { getNumbers } from './utils';
 
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import React, { useEffect, useState } from 'react';
+import { BrowserRouter, Route, Routes, useNavigate } from 'react-router-dom';
+
+import { getNumbers, maybeParseInt } from './utils';
+
+import { useUrlQuery } from './hooks/useUrlQuery';
+
+import { Pagination } from './components/Pagination/Pagination';
+
 const items = getNumbers(1, 42).map(n => `Item ${n}`);
 
-export const App: React.FC = () => {
+const RoutedApp: React.FC = () => {
+  const query = useUrlQuery();
+  const navigate = useNavigate();
+
+  const [perPage, setPerPage] = useState(
+    maybeParseInt(query.get('perPage')) ?? 5,
+  );
+  const [currentPage, setCurrentPage] = useState(
+    maybeParseInt(query.get('page')) ?? 1,
+  );
+
+  useEffect(() => {
+    if (!query.get('page') || !query.get('perPage')) {
```

Comment:
The condition here is checking if 'page' or 'perPage' are not present in the URL, but it's not checking for valid values. Consider adding validation for the query parameters to ensure they are positive integers before navigating to the updated URL.